contrib/hyperscan/matching/input_matchers/test/matcher_test.cc

```diff
@@ -48,6 +48,20 @@ TEST(ThreadLocalTest, RaceScratchCreation) {
   }
 }
 
+// Verify that even if thread local is not initialized, matcher can work without crashes.
+TEST(ThreadLocalTest, NotInitialized) {
+  std::vector<const char*> expressions{"^/asdf/.+"};
+  std::vector<unsigned int> flags{0};
+  std::vector<unsigned int> ids{0};
+
+  ThreadLocal::InstanceImpl instance;
+  Matcher matcher(expressions, flags, ids, instance, false);
+  instance.shutdownThread();
```

Comment:
A question, how the ThreadLocal can be not initialized? Since we already invoke Matcher's constructor with a ThreadLocal instance.  Is that due to the thread doesn't register yet?

Reply:
Yes. E.g., source/server/server.cc#L465 creates the regex engine, L471 creates matchers which may include regex matchers, while the thread will keep not registered until L638.
So any using of regex matchers before L638 will allocate temporary scratch itself for matching and use scratch in TLS after L638.

Reply:
discussed with @zhxie offline, we should update the tls after the new thread is registered, then the new thread won't keep using the temp scratch all the time.